arbor/include/arbor/serdes.hpp

```diff
@@ -0,0 +1,476 @@
+#pragma once
+
+#include <optional>
+#include <string_view>
+#include <string>
+#include <vector>
+#include <memory>
+#include <map>
+#include <unordered_map>
+#include <array>
+#include <stdexcept>
+
+#include <arbor/export.hpp>
+
+namespace arb {
+
+// NOTE: Cannot use arbexcept since it include common_type, which includes this. Circular include!
+struct ARB_SYMBOL_VISIBLE serdes_error: std::runtime_error {
+    serdes_error(std::string_view w): std::runtime_error{std::string{w}} {}
+};
+
+struct ARB_SYMBOL_VISIBLE illegal_key_type: serdes_error {
+    illegal_key_type(): serdes_error{"SerDes keys must be an integral or string-like type."} {}
+};
+
+// Handling keys
+using key_type = std::string;
+
+template <typename K>
+key_type to_key(K&& key) {
+    using T = std::decay_t<std::remove_cv_t<std::remove_reference_t<K>>>;
+    if constexpr (std::is_same_v<std::string, T>) {
+        return key;
+    }
+    else if constexpr (std::is_same_v<char*, T>) {
+        return std::string{key};
+    }
+    else {
+        return std::to_string(key);
+    }
+}
+
+template <typename K>
+void from_key(K& key, const key_type& k) {
+    using T = std::decay_t<std::remove_cv_t<std::remove_reference_t<K>>>;
+    if constexpr (std::is_same_v<std::string, T>) {
+        key = k;
+    }
+    else if constexpr (std::is_same_v<char*, T>) {
+        key = std::string{k};
```

Comment:
If `K` is a stringy type, it might be a (possibly  `const`) `char[]`, `char*`, or a `std::string`, 
or a reference to any of those.
After `decay` and stripping `&`, `const`, and `volatile` we have `char*` left (or `std::string`), no?

Reply:
agreed. My question was about the assignment. The lhs seems to be a `char*`. I was expecting something like this: `key = k.c_str();`. Do I have it backwards?

Reply:
Ah, good catch! Although, that would be incorrect, since assume the original type is a `char[5]` and we try
to write `foobarbaz` into it. Not only won't this compile, it would be really Bad if it did, which it will for `char*` where it will potentially overwrite unowned memory. BAD. So, I'd conclude we don't allow converting to `char*`/`char[]`, due to security issues. Conversvion _from_ can be more lenient.

Reply:
Other options will run into problems, too. 
- Who will have deallocate the `char*`?
- We could not deallocate, but that leaks
- Also, we could have a pointer into nothingness since the original string might disappear. 

Thus, we can only copy out _values_.

Reply:
I concur